File: tests/algorithms/x/Clojure/data_structures/arrays/median_two_array.clj
(ns main (:refer-clojure :exclude [sortFloats find_median_sorted_arrays]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare sortFloats find_median_sorted_arrays)

(def ^:dynamic find_median_sorted_arrays_i nil)

(def ^:dynamic find_median_sorted_arrays_j nil)

(def ^:dynamic find_median_sorted_arrays_merged nil)

(def ^:dynamic find_median_sorted_arrays_middle1 nil)

(def ^:dynamic find_median_sorted_arrays_middle2 nil)

(def ^:dynamic find_median_sorted_arrays_sorted nil)

(def ^:dynamic find_median_sorted_arrays_total nil)

(def ^:dynamic sortFloats_arr nil)

(def ^:dynamic sortFloats_i nil)

(def ^:dynamic sortFloats_j nil)

(def ^:dynamic sortFloats_t nil)

(defn sortFloats [sortFloats_xs]
  (binding [sortFloats_arr nil sortFloats_i nil sortFloats_j nil sortFloats_t nil] (try (do (set! sortFloats_arr sortFloats_xs) (set! sortFloats_i 0) (while (< sortFloats_i (count sortFloats_arr)) (do (set! sortFloats_j 0) (while (< sortFloats_j (- (count sortFloats_arr) 1)) (do (when (> (nth sortFloats_arr sortFloats_j) (nth sortFloats_arr (+ sortFloats_j 1))) (do (set! sortFloats_t (nth sortFloats_arr sortFloats_j)) (set! sortFloats_arr (assoc sortFloats_arr sortFloats_j (nth sortFloats_arr (+ sortFloats_j 1)))) (set! sortFloats_arr (assoc sortFloats_arr (+ sortFloats_j 1) sortFloats_t)))) (set! sortFloats_j (+ sortFloats_j 1)))) (set! sortFloats_i (+ sortFloats_i 1)))) (throw (ex-info "return" {:v sortFloats_arr}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn find_median_sorted_arrays [find_median_sorted_arrays_nums1 find_median_sorted_arrays_nums2]
  (binding [find_median_sorted_arrays_i nil find_median_sorted_arrays_j nil find_median_sorted_arrays_merged nil find_median_sorted_arrays_middle1 nil find_median_sorted_arrays_middle2 nil find_median_sorted_arrays_sorted nil find_median_sorted_arrays_total nil] (try (do (when (and (= (count find_median_sorted_arrays_nums1) 0) (= (count find_median_sorted_arrays_nums2) 0)) (throw (Exception. "Both input arrays are empty."))) (set! find_median_sorted_arrays_merged []) (set! find_median_sorted_arrays_i 0) (while (< find_median_sorted_arrays_i (count find_median_sorted_arrays_nums1)) (do (set! find_median_sorted_arrays_merged (conj find_median_sorted_arrays_merged (nth find_median_sorted_arrays_nums1 find_median_sorted_arrays_i))) (set! find_median_sorted_arrays_i (+ find_median_sorted_arrays_i 1)))) (set! find_median_sorted_arrays_j 0) (while (< find_median_sorted_arrays_j (count find_median_sorted_arrays_nums2)) (do (set! find_median_sorted_arrays_merged (conj find_median_sorted_arrays_merged (nth find_median_sorted_arrays_nums2 find_median_sorted_arrays_j))) (set! find_median_sorted_arrays_j (+ find_median_sorted_arrays_j 1)))) (set! find_median_sorted_arrays_sorted (sortFloats find_median_sorted_arrays_merged)) (set! find_median_sorted_arrays_total (count find_median_sorted_arrays_sorted)) (when (= (mod find_median_sorted_arrays_total 2) 1) (throw (ex-info "return" {:v (nth find_median_sorted_arrays_sorted (quot find_median_sorted_arrays_total 2))}))) (set! find_median_sorted_arrays_middle1 (nth find_median_sorted_arrays_sorted (- (quot find_median_sorted_arrays_total 2) 1))) (set! find_median_sorted_arrays_middle2 (nth find_median_sorted_arrays_sorted (quot find_median_sorted_arrays_total 2))) (throw (ex-info "return" {:v (/ (+ find_median_sorted_arrays_middle1 find_median_sorted_arrays_middle2) 2.0)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (find_median_sorted_arrays [1.0 3.0] [2.0]))
      (println (find_median_sorted_arrays [1.0 2.0] [3.0 4.0]))
      (println (find_median_sorted_arrays [0.0 0.0] [0.0 0.0]))
      (println (find_median_sorted_arrays [] [1.0]))
      (println (find_median_sorted_arrays [(- 1000.0)] [1000.0]))
      (println (find_median_sorted_arrays [(- 1.1) (- 2.2)] [(- 3.3) (- 4.4)]))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Pascal/data_structures/arrays/median_two_array.pas
{$mode objfpc}
program Main;
uses SysUtils;
type RealArray = array of real;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  nums2: RealArray;
  nums1: RealArray;
  xs: RealArray;
function sortFloats(xs: RealArray): RealArray; forward;
function find_median_sorted_arrays(nums1: RealArray; nums2: RealArray): real; forward;
function sortFloats(xs: RealArray): RealArray;
var
  sortFloats_arr: array of real;
  sortFloats_i: integer;
  sortFloats_j: integer;
  sortFloats_t: real;
begin
  sortFloats_arr := xs;
  sortFloats_i := 0;
  while sortFloats_i < Length(sortFloats_arr) do begin
  sortFloats_j := 0;
  while sortFloats_j < (Length(sortFloats_arr) - 1) do begin
  if sortFloats_arr[sortFloats_j] > sortFloats_arr[sortFloats_j + 1] then begin
  sortFloats_t := sortFloats_arr[sortFloats_j];
  sortFloats_arr[sortFloats_j] := sortFloats_arr[sortFloats_j + 1];
  sortFloats_arr[sortFloats_j + 1] := sortFloats_t;
end;
  sortFloats_j := sortFloats_j + 1;
end;
  sortFloats_i := sortFloats_i + 1;
end;
  exit(sortFloats_arr);
end;
function find_median_sorted_arrays(nums1: RealArray; nums2: RealArray): real;
var
  find_median_sorted_arrays_merged: array of real;
  find_median_sorted_arrays_i: integer;
  find_median_sorted_arrays_j: integer;
  find_median_sorted_arrays_sorted: RealArray;
  find_median_sorted_arrays_total: integer;
  find_median_sorted_arrays_middle1: real;
  find_median_sorted_arrays_middle2: real;
begin
  if (Length(nums1) = 0) and (Length(nums2) = 0) then begin
  panic('Both input arrays are empty.');
end;
  find_median_sorted_arrays_merged := [];
  find_median_sorted_arrays_i := 0;
  while find_median_sorted_arrays_i < Length(nums1) do begin
  find_median_sorted_arrays_merged := concat(find_median_sorted_arrays_merged, [nums1[find_median_sorted_arrays_i]]);
  find_median_sorted_arrays_i := find_median_sorted_arrays_i + 1;
end;
  find_median_sorted_arrays_j := 0;
  while find_median_sorted_arrays_j < Length(nums2) do begin
  find_median_sorted_arrays_merged := concat(find_median_sorted_arrays_merged, [nums2[find_median_sorted_arrays_j]]);
  find_median_sorted_arrays_j := find_median_sorted_arrays_j + 1;
end;
  find_median_sorted_arrays_sorted := sortFloats(find_median_sorted_arrays_merged);
  find_median_sorted_arrays_total := Length(find_median_sorted_arrays_sorted);
  if (find_median_sorted_arrays_total mod 2) = 1 then begin
  exit(find_median_sorted_arrays_sorted[find_median_sorted_arrays_total div 2]);
end;
  find_median_sorted_arrays_middle1 := find_median_sorted_arrays_sorted[(find_median_sorted_arrays_total div 2) - 1];
  find_median_sorted_arrays_middle2 := find_median_sorted_arrays_sorted[find_median_sorted_arrays_total div 2];
  exit((find_median_sorted_arrays_middle1 + find_median_sorted_arrays_middle2) / 2);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(find_median_sorted_arrays([1, 3], [2]));
  writeln(find_median_sorted_arrays([1, 2], [3, 4]));
  writeln(find_median_sorted_arrays([0, 0], [0, 0]));
  writeln(find_median_sorted_arrays(RealArray([]), [1]));
  writeln(find_median_sorted_arrays([-1000], [1000]));
  writeln(find_median_sorted_arrays([-1.1, -2.2], [-3.3, -4.4]));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
